structured_prediction_baselines/modules/sampler/sg_spen.py

```diff
@@ -0,0 +1,320 @@
+                    samples = self._get_samples(y_inf, self.n_samples)
+                    sample_pairs = list(itertools.permutations(samples.transpose(0, 1), 2))
+                    y_hat, y_hat_extra = tuple(map(torch.stack, zip(*sample_pairs)))
+                    y_hat = torch.transpose(y_hat, 0, 1)
+                    y_hat_extra = torch.transpose(y_hat_extra, 0, 1)
+                    loss_value = self.update(
+                        y_hat, y_hat_extra, buffer, loss_fn
+                    )
+
+                    loss_value = torch.mean(loss_value)
+                    loss_values.append(float(loss_value))
+
+                    step_number += 1
+                self._metrics[self.name + '_loss'] = np.mean(loss_values)
+                self._total_loss += np.mean(loss_values)
+                self._num_batches += 1
+
+            return (
+                y_hat.detach().clone(),
+                y_hat_extra.detach().clone()
+            )
+
+    def update(
+        self,
+        y_hat: torch.Tensor,
+        y_cost_aug: torch.Tensor,
+        buffer: Dict,
+        loss_fn: Callable[[torch.Tensor, torch.Tensor, Dict], torch.Tensor],
+    ) -> torch.Tensor:
+        """
+        Returns:
+            loss: loss value at the previous point (unreduced)
+        """
+        loss = loss_fn(y_hat, y_cost_aug, buffer)
+
+        if self.optimizer:
+            loss.backward()  # type:ignore
+            self.optimizer.step()
+
+        return loss
+
+    def _get_samples(
+        self,
+        y_pred: torch.Tensor,
+        n_samples: int
+    ) -> torch.Tensor:
+        if not self.is_normalized:
+            y_pred = torch.sigmoid(y_pred)
+
+        p = y_pred.squeeze(1)  # (batch, num_labels)
+        distribution = torch.distributions.Bernoulli(probs=p)
+        samples = torch.transpose(distribution.sample([n_samples]), 0, 1)
+        return samples
+
+    def _get_values(
+        self,
+        x: Any,
+        labels: torch.Tensor,
+        buffer: Dict,
+        **kwargs: Any
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+
+        y_inf: torch.Tensor = self.inference_nn(
+            x, buffer
+        ).unsqueeze(
+            1
+        )  # (batch_size, 1, ...) unormalized
+        # inference_nn is TaskNN so it will output tensor of shape (batch, ...)
+        # hence the unsqueeze
+
+        if self.cost_augmented_layer is not None:
+            y_cost_aug = self.cost_augmented_layer(
+                torch.cat(
+                    (
+                        y_inf.squeeze(1),
+                        labels.to(dtype=y_inf.dtype).squeeze(1),
+                    ),
+                    dim=-1,
+                ),
+                buffer,
+            ).unsqueeze(
+                1
+            )  # (batch_size,1, ...)
+        else:
+            y_cost_aug = None
+
+        return y_inf, y_cost_aug
+
+    def get_metrics(self, reset: bool = False):
```

Comment:
You don't need to implement `get_metrics()` on sampler anymore. Have a look at the wiki of the repo.